File: HexChat/highlight.py
import hexchat

__module_name__ = 'Highlight Logger'
__module_version__ = '2'
__module_description__ = 'Prints highlights to another tab'
# Forked from Arnavion's

TAB_NAME = '(highlights)'

def find_highlighttab():
	context = hexchat.find_context(channel=TAB_NAME)
	if context == None: # Create a new one in the background
		newtofront = hexchat.get_prefs('gui_tab_newtofront')

		hexchat.command('set -quiet gui_tab_newtofront 0')
		hexchat.command('newserver -noconnect {0}'.format(TAB_NAME))
		hexchat.command('set -quiet gui_tab_newtofront {}'.format(newtofront))

		return hexchat.find_context(channel=TAB_NAME)
	else:
		return context

def highlight_callback(word, word_eol, user_data):
	
	word = [(word[i] if len(word) > i else '') for i in range(4)]
	
	highlight_context = find_highlighttab()
	
	channel = hexchat.get_info('channel')
	
	if user_data == 'Channel Msg Hilight':
		highlight_context.prnt('\00322{0}\t\00318<{4}{3}{1}>\017 {2}'.format(channel, *word))
	elif user_data == 'Channel Action Hilight':
		highlight_context.prnt('\00322{0}\t\002\00318{4}{3}{1}\017 {2}'.format(channel, *word))	

	# Can't easily intelligently manage color so lets just ignore it
	highlight_context.command('gui color 0')
	
	return hexchat.EAT_NONE

hexchat.hook_print('Channel Msg Hilight', highlight_callback, 'Channel Msg Hilight')
hexchat.hook_print('Channel Action Hilight', highlight_callback, 'Channel Action Hilight')

File: HexChat/highlight.lua
-- SPDX-License-Identifier: MIT
hexchat.register('Highlights', '1', 'Prints highlights to another tab')

local TAB_NAME = '(highlights)'
local OPEN_PER_SERVER = false

local function find_highlighttab ()
	local network = nil
	if OPEN_PER_SERVER then
		network = hexchat.get_info('network')
	end
	local ctx = hexchat.find_context(network, TAB_NAME)
	if not ctx then
		if OPEN_PER_SERVER then
			hexchat.command('query -nofocus ' .. TAB_NAME)
		else
			local newtofront = hexchat.prefs['gui_tab_newtofront']
			hexchat.command('set -quiet gui_tab_newtofront off')
			hexchat.command('newserver -noconnect ' .. TAB_NAME)
			hexchat.command('set -quiet gui_tab_newtofront ' .. tostring(newtofront))
		end

		return hexchat.find_context(network, TAB_NAME)
	end

	return ctx
end

local function on_highlight (args, event_type)
	local channel = hexchat.get_info('channel')
	local highlight_context = find_highlighttab()

	local format
	if event_type == 'Channel Msg Hilight' then
		format = '\00322%s\t\00318<%s%s%s>\015 %s'
	elseif event_type == 'Channel Action Hilight' then
		format = '\00322%s\t\002\00318%s%s%s\015 %s'
	end

	highlight_context:print(string.format(format, channel,
	                       args[3] or '', args[4] or '', hexchat.strip(args[1]), args[2]))
	highlight_context:command('gui color 0') -- Ignore colors
end

for _, event in ipairs({'Channel Msg Hilight', 'Channel Action Hilight'}) do
	hexchat.hook_print(event, function (args)
		return on_highlight(args, event)
	end, hexchat.PRI_LOW)
end
